colcon_cargo/task/cargo/test.py

```diff
@@ -31,29 +38,98 @@ async def test(self, *, additional_hooks=None):  # noqa: D102
         logger.info(
             "Testing Cargo package in '{args.path}'".format_map(locals()))
 
-        assert os.path.exists(args.build_base)
+        assert os.path.exists(args.build_base), \
+            'Has this package been built before?'
 
-        test_results_path = os.path.join(args.build_base, 'test_results')
-        os.makedirs(test_results_path, exist_ok=True)
+        test_results_path = os.path.join(args.build_base, 'cargo_test.xml')
 
         try:
             env = await get_command_environment(
                 'test', args.build_base, self.context.dependencies)
         except RuntimeError as e:
+            # TODO(luca) log this as error in the test result file
             logger.error(str(e))
             return 1
 
+        # Disable color to avoid escape sequences in test result file
+        env['NO_COLOR'] = '1'
+
         if CARGO_EXECUTABLE is None:
+            # TODO(luca) log this as error in the test result file
             raise RuntimeError("Could not find 'cargo' executable")
```

Comment:
Fair point, I just left it unchanged to avoid touching up too much. This behavior is also what the `build` task does https://github.com/colcon/colcon-cargo/blob/9fdb14f6b1149d3d3b45c3ab6a90be5491861f54/colcon_cargo/task/cargo/build.py#L47-L69 so imho if we change it here we should also change it there for consistency, and then we would have a PR about `cargo test` going to touch on the `build` task!
What the TODO documents is that I was pondering whether we should create a test-result file for this case or just exit early with an error. I suspect we should still create a test result file hence I noted that down as a TODO.

For reference making it a return code would produce this result:

![image](https://github.com/user-attachments/assets/66f36dbb-a450-4c87-8980-0bac53ea75fa)

While currently (`RuntimeError`) this is the output:

![image](https://github.com/user-attachments/assets/78ab1972-45a4-45ce-a176-281b1680d457)

My recommendation here would be to decide which looks best and do a followup PR that fixes it for both the build and test task. It will be a choice between verbosity (raise exception, get full backtrace) and simplicity (return error code, simple and concise error message). But as a spoiler I do agree with you that the concise error code looks a lot better, most people probably don't want a backtrace of colcon's task spawning

Reply:
Removed the TODO https://github.com/colcon/colcon-cargo/pull/39/commits/b3ef3cde270223f2f5550b7e6907af6553ba2e55